Druid/DRPiDeadBranchElimination.class.st

```diff
@@ -0,0 +1,74 @@
+Class {
+	#name : 'DRPiDeadBranchElimination',
+	#superclass : 'DROptimisation',
+	#instVars : [
+		'worklist',
+		'visited'
+	],
+	#category : 'Druid-Optimizations',
+	#package : 'Druid',
+	#tag : 'Optimizations'
+}
+
+{ #category : 'accessing' }
+DRPiDeadBranchElimination >> applyTo: cfg [
+
+	| deadEdges |
+	worklist add: (DREdge
+			 withSource: cfg initialBasicBlock
+			 andDestination: cfg initialBasicBlock).
+	"The source doesn't matter in the initial edge, it'll be never be dead anyways"
+
+	deadEdges := self computeDeadeEdges.
+
+	deadEdges ifNotEmpty: [
+		self removeUnnecessaryConditionalJumps: deadEdges.
+		cfg applyOptimisation: DRDeadBlockElimination new ]
+]
+
+{ #category : 'accessing' }
+DRPiDeadBranchElimination >> computeDeadeEdges [
+
+	| deadEdges |
+	deadEdges := OrderedCollection new.
+
+	[ worklist isEmpty ] whileFalse: [
+		| currentEdge |
+		currentEdge := worklist removeFirst.
+		(self isReachable: currentEdge destination)
+			ifTrue: [ self updateWorklistFrom: currentEdge destination ]
+			ifFalse: [ deadEdges add: currentEdge ] ].
+		"TODO if a block is unreachable, then its sibling must be reachable,
+		so we can remove it from the worklist"
+	^ deadEdges
+]
```

Comment:
I think that we don't need the edges here. We can identify the piNodes with _dead (full) constraint_ (based on previous constraints too), then those piNodes should have only one predecessor (TOCHECK), so we can just disconnect the basicBlock from the predecessor.

Reply:
https://github.com/Alamvic/druid/issues/185